### PR TITLE
feat: Add models for connection, sync, sync_run, sync_log, sync_write_record and sync_read_record

### PR DESCRIPTION
**Pull request by karthik-sivadas**
## Release Notes

- **New Features**
  - Introduced new models for managing synchronization: `Connection`, `Sync`, `SyncLog`, `SyncReadRecord`, `SyncRun`, and `SyncWriteRecord`.
  - Added functionality for logging synchronization activities and managing sync states.
  - Implemented a new database schema to support connections and sync operations.

- **Bug Fixes**
  - Enhanced validations to ensure data integrity across new models.

- **Documentation**
  - Added comprehensive test coverage for new models, ensuring expected behaviors and edge cases are handled.

- **Chores**
  - Updated dependencies in the Gemfile to include `deepsort` and `json-schema`.

### CODE REVIEW

File: platform/db/migrate/20241001105047_create_sync_read_records.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    create_table :sync_read_records do |t|
+      t.references :sync_run, null: false, foreign_key: true
+      t.references :sync, null: false, foreign_key: true
+      t.jsonb :data, null: false
+      t.string :signature, null: false
+
+      t.timestamps
+    end
+  end
+end
```
[the remark this answers is not in the document]
Comment (karthik-sivadas) [reply platform/db/migrate/20241001105047_create_sync_read_records.rb:14]: There are scenarios when we need to duplicate data. In those cases, we will need to include duplication. So, it is not advised to have unique index based on above parameters

File: platform/db/migrate/20241001105046_create_sync_runs.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+class CreateSyncRuns < ActiveRecord::Migration[7.1]
+  # rubocop:disable Metrics/MethodLength
+  def change
+    create_table :sync_runs do |t|
+      t.references :sync, null: false, foreign_key: true
+      t.integer :status, null: false, default: 0
```
[the remark this answers is not in the document]
Comment (karthik-sivadas) [reply platform/db/migrate/20241001105046_create_sync_runs.rb:8]: We already have an enum in place. This suggestion needs to be corrected.

File: platform/app/models/sync_read_record.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+class SyncReadRecord < ApplicationRecord
+  belongs_to :sync_run
+  belongs_to :sync
+
+  validates :data, presence: true
+
+  before_validation :generate_signature
```
[the remark this answers is not in the document]
Comment (karthik-sivadas) [reply platform/app/models/sync_read_record.rb:9]: Explained above why this is not needed
